File: programmers/lv2/게임맵최단거리.py
from collections import deque

def solution(maps):
    q = deque()
    x, y = len(maps), len(maps[0])
    visit = [[0] * y for _ in range(x)]
    cost = [[1000000] * y for _ in range(x)]
    cost[0][0] = 0
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    q.append([0,0])
    
    while(q):
        currX, currY = q.popleft()
        if visit[currX][currY] == 0:
            visit[currX][currY] = 1
            for i in range(4):
                nextX, nextY = currX+dx[i], currY+dy[i]
                if (0 <= nextX < x) and (0 <= nextY < y):
                    if maps[nextX][nextY] == 1:
                        cost[nextX][nextY] = min(cost[nextX][nextY], cost[currX][currY] + 1)
                        q.append([nextX, nextY])
    
    answer = cost[x - 1][y - 1] + 1 if visit[x - 1][y - 1] == 1 else -1
    return answer

solution([[1,0,1,1,1],[1,0,1,0,1],[1,0,1,1,1],[1,1,1,0,1],[0,0,0,0,1]])

File: programmers/lv2/게임맵최단거리.js
function solution(maps) {
  let answer = -1;
  let dx = [-1, 1, 0, 0];
  let dy = [0, 0, -1, 1];
  let q = [];

  let visit = [];
  for (let i = 0; i < maps.length; i++) {
    let row = [];
    for (let j = 0; j < maps[i].length; j++) {
      row.push(0);
    }
    visit.push(row);
  }

  q.push([0, 0]);
  visit[0][0] = 1;

  while (q.length > 0) {
    let pos = q.shift();
    // console.log('방문', pos);
    if (pos[0] === maps.length - 1 && pos[1] === maps[0].length - 1) {
      answer = visit[visit.length - 1][visit[0].length - 1];
    }
    for (let i = 0; i < 4; i++) {
      let mx = pos[0] + dx[i];
      let my = pos[1] + dy[i];
      if (0 <= mx && mx < maps.length && 0 <= my && my < maps[0].length) {
        if (maps[mx][my] === 1 && visit[mx][my] === 0) {
          visit[mx][my] = visit[pos[0]][pos[1]] + 1;
          q.push([mx, my]);
        }
      }
    }
  }
  return answer;
}
